File: CLDemo-Swift/CLDemo-Swift/CLRecorder/CLRecorder.m
//
//  CLRecorder.m
//  CLDemo
//
//  Created by JmoVxia on 2020/6/24.
//  Copyright © 2020 JmoVxia. All rights reserved.
//

#import "CLRecorder.h"
#import <AVFoundation/AVFoundation.h>
#import "CLMp3Encoder.h"

#define INPUT_BUS 1

@interface CLRecorder ()

@property (nonatomic, assign) AudioUnit audioUnit;

@property (nonatomic, strong) CLMp3Encoder *mp3Encoder;

@property (nonatomic, copy) NSString *mp3Path;

@property (nonatomic, strong) NSLock *lock;

@property (nonatomic, strong) NSFileHandle * handle;

@property (nonatomic, assign) BOOL otherAudioPlaying;

@property (nonatomic, assign) NSUInteger lastSecond;

@end

@implementation CLRecorder

static OSStatus RecordCallback(void *inRefCon,
                               AudioUnitRenderActionFlags *ioActionFlags,
                               const AudioTimeStamp *inTimeStamp,
                               UInt32 inBusNumber,
                               UInt32 inNumberFrames,
                               AudioBufferList *ioData)
{
    AudioBufferList bufferList;
    bufferList.mNumberBuffers = 1;
    bufferList.mBuffers[0].mNumberChannels = 1;
    bufferList.mBuffers[0].mData = NULL;
    bufferList.mBuffers[0].mDataByteSize = 0;
    
    CLRecorder *recorder = (__bridge CLRecorder *)(inRefCon);
    AudioUnitRender(recorder.audioUnit, ioActionFlags, inTimeStamp, inBusNumber, inNumberFrames, &bufferList);
    [recorder.mp3Encoder processAudioBufferList:bufferList];
    CGFloat audioDuration = recorder.audioDuration;
    NSUInteger second = (NSUInteger)floor(audioDuration);
    if (recorder.lastSecond != second) {
        recorder.lastSecond = second;
        if (recorder.durationCallback) {
            dispatch_async(dispatch_get_main_queue(), ^{
                recorder.durationCallback(second);
            });
        }
    }
    return noErr;
}
- (instancetype)init {
    self = [super init];
    if (self) {
        [self initFold];
        [self.mp3Encoder run];
        [self initRemoteIO];
    }
    return self;
}
- (void)initFold {
    NSString * recorderFoldPath = [NSHomeDirectory() stringByAppendingString:@"/Documents/CLChatRecorder"];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    BOOL isDir = NO;
    if (![fileManager fileExistsAtPath:recorderFoldPath isDirectory:&isDir] ) {//如果文件夹不存在
        [fileManager createDirectoryAtPath:recorderFoldPath withIntermediateDirectories:YES attributes:nil error:nil];
    }
}
- (void)initRemoteIO {
    AudioUnitInitialize(self.audioUnit);
    [self initBuffer];
    [self initAudioComponent];
    [self initFormat];
    [self initAudioProperty];
    [self initRecordeCallback];
}
- (void)initBuffer {
    UInt32 flag = 0;
    AudioUnitSetProperty(self.audioUnit,
                         kAudioUnitProperty_ShouldAllocateBuffer,
                         kAudioUnitScope_Output,
                         INPUT_BUS,
                         &flag,
                         sizeof(flag));
}
- (void)initAudioComponent {
    AudioComponentDescription audioDesc;
    audioDesc.componentType = kAudioUnitType_Output;
    audioDesc.componentSubType = kAudioUnitSubType_RemoteIO;
    audioDesc.componentManufacturer = kAudioUnitManufacturer_Apple;
    audioDesc.componentFlags = 0;
    audioDesc.componentFlagsMask = 0;
    
    AudioComponent inputComponent = AudioComponentFindNext(NULL, &audioDesc);
    AudioComponentInstanceNew(inputComponent, &_audioUnit);
}
- (void)initFormat {
    AudioStreamBasicDescription audioFormat;
    audioFormat.mSampleRate = 44100;
    audioFormat.mFormatID = kAudioFormatLinearPCM;
    audioFormat.mFormatFlags = kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsNonInterleaved | kAudioFormatFlagIsPacked;
    audioFormat.mFramesPerPacket = 1;
    audioFormat.mChannelsPerFrame = 1;
    audioFormat.mBitsPerChannel = 16;
    audioFormat.mBytesPerFrame = (audioFormat.mBitsPerChannel / 8) * audioFormat.mChannelsPerFrame;
    audioFormat.mBytesPerPacket = audioFormat.mBytesPerFrame;;
    
    AudioUnitSetProperty(self.audioUnit,
                         kAudioUnitProperty_StreamFormat,
                         kAudioUnitScope_Output,
                         INPUT_BUS,
                         &audioFormat,
                         sizeof(audioFormat));
}
- (void)initAudioProperty {
    UInt32 flag = 1;
    AudioUnitSetProperty(self.audioUnit,
                         kAudioOutputUnitProperty_EnableIO,
                         kAudioUnitScope_Input,
                         INPUT_BUS,
                         &flag,
                         sizeof(flag));
}
- (void)initRecordeCallback {
    AURenderCallbackStruct recordCallback;
    recordCallback.inputProc = RecordCallback;
    recordCallback.inputProcRefCon = (__bridge void *)self;
    AudioUnitSetProperty(self.audioUnit,
                         kAudioOutputUnitProperty_SetInputCallback,
                         kAudioUnitScope_Global,
                         INPUT_BUS,
                         &recordCallback,
                         sizeof(recordCallback));
}
- (void)startRecorder {
    self.lastSecond = 0;
    if (self.durationCallback) {
        dispatch_async(dispatch_get_main_queue(), ^{
            self.durationCallback(0);
        });
    }
    NSString *foldPath = [NSHomeDirectory() stringByAppendingFormat:@"/Documents/Record"];
    if (![[NSFileManager defaultManager] fileExistsAtPath:foldPath]) {
        [[NSFileManager defaultManager] createDirectoryAtPath:foldPath withIntermediateDirectories:true attributes:nil error:nil];
    }
    self.mp3Path = [foldPath stringByAppendingFormat:@"/%@.mp3", [self currentTime]];
    if (![[NSFileManager defaultManager] fileExistsAtPath: self.mp3Path]) {
        [[NSFileManager defaultManager] createFileAtPath: self.mp3Path contents:nil attributes:nil];
    }
    NSError *error;
    self.handle = [NSFileHandle fileHandleForWritingToURL:[NSURL fileURLWithPath: self.mp3Path] error:&error];
    if (!error) {
        [self activeAudioSession];
        OSStatus status = AudioOutputUnitStart(self.audioUnit);
        if (status != noErr) {
            NSLog(@"startRecorder error %d", status);
        }
    }else {
        NSLog(@"error: %@", error);
    }
}
- (void)cancelRecorder {
    OSStatus status = AudioOutputUnitStop(self.audioUnit);
    if (status != noErr) {
        NSLog(@"stopRecorder error: %d", status);
    }
    [self resumeActiveAudioSession];
    [self removeFile];
}
- (void)stopRecorder {
    OSStatus status = AudioOutputUnitStop(self.audioUnit);
    [self resumeActiveAudioSession];
    if (status != noErr) {
        NSLog(@"stopRecorder error: %d", status);
    }else {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (self.finishCallBack) {
                self.finishCallBack(self.audioDuration, [[NSData alloc] initWithContentsOfFile:self.mp3Path]);
            }
        });
    }
}
- (void)removeFile {
    if ([[NSFileManager defaultManager] fileExistsAtPath: self.mp3Path]) {
        [[NSFileManager defaultManager] removeItemAtPath:self.mp3Path error:nil];
        self.mp3Path = nil;
    }
}
- (void)activeAudioSession {
    self.otherAudioPlaying = [AVAudioSession sharedInstance].otherAudioPlaying;
    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
    [audioSession setCategory:AVAudioSessionCategoryRecord error:nil];
    [audioSession setPreferredSampleRate:44100 error:nil];
    [audioSession setPreferredInputNumberOfChannels:1 error:nil];
    [audioSession setPreferredIOBufferDuration:0.023 error:nil];
    [audioSession setActive:YES error:nil];
}
- (void)resumeActiveAudioSession {
    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
    if (self.otherAudioPlaying) {
        [audioSession setActive:YES withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:nil];
    }else {
        [audioSession setCategory:AVAudioSessionCategoryPlayback error:nil];
        [audioSession setActive:YES error:nil];
    }
}
- (void)writeData:(NSData *)data {
    [self.lock lock];
    [self.handle seekToEndOfFile];
    [self.handle writeData:data];
    [self.lock unlock];
}
- (NSString *)currentTime {
    NSTimeInterval time = [[NSDate date] timeIntervalSince1970];
    return [NSString stringWithFormat:@"%ld",(long)((CGFloat)time * 1000000)];
}
- (CGFloat)audioDuration {
    if (self.mp3Path.length > 0) {
        NSURL *url = [NSURL fileURLWithPath:self.mp3Path];
        AVURLAsset* audioAsset = [AVURLAsset URLAssetWithURL:url options:nil];
        CMTime audioDuration = audioAsset.duration;
        return CMTimeGetSeconds(audioDuration);
    }else {
        return 0.0;
    }
}
- (void)dealloc {
    [self.mp3Encoder stop];
    AudioUnitUninitialize(self.audioUnit);
}
- (CLMp3Encoder *)mp3Encoder {
    if (!_mp3Encoder) {
        _mp3Encoder = [[CLMp3Encoder alloc] init];
        _mp3Encoder.inputSampleRate = 44100;
        _mp3Encoder.outputSampleRate = 44100;
        _mp3Encoder.outputChannelsPerFrame = 1;
        _mp3Encoder.bitRate = 16;
        _mp3Encoder.quality = 9;
        __weak __typeof(self) weakSelf = self;
        _mp3Encoder.processingEncodedData = ^(NSData * _Nonnull mp3Data) {
            __typeof(&*weakSelf) strongSelf = weakSelf;
            [strongSelf writeData:mp3Data];
        };
    }
    return _mp3Encoder;
}
- (NSLock *)lock {
    if (!_lock) {
        _lock = [[NSLock alloc] init];
    }
    return _lock;
}
@end

File: CLDemo-Swift/CLDemo-Swift/CLRecorder/CLRecorder.swift
//
//  CLRecorder.swift
//  CKD
//
//  Created by Chen JmoVxia on 2024/5/28.
//  Copyright © 2024 JmoVxia. All rights reserved.
//

import AudioToolbox
import AVFoundation
import Foundation

class CLRecorder: NSObject {
    override init() {
        super.init()
        initFolder()
        initRemoteIO()
        mp3Encoder.run()
    }

    deinit {
        mp3Encoder.stop()
        guard let audioUnit else { return }
        AudioUnitUninitialize(audioUnit)
    }

    private lazy var mp3Encoder: CLMp3Encoder = {
        let encoder = CLMp3Encoder()
        encoder.inputSampleRate = 44100
        encoder.outputSampleRate = 44100
        encoder.outputChannelsPerFrame = 1
        encoder.bitRate = 16
        encoder.quality = 9
        encoder.processingEncodedData = { [weak self] mp3Data in
            guard let self else { return }
            writeData(mp3Data)
        }
        return encoder
    }()

    private var mp3Path: String = ""

    private let lock = NSLock()

    private var audioUnit: AudioUnit?

    private var fileHandle: FileHandle?

    private var otherAudioPlaying = false

    private var lastSecond = Int.zero

    var durationCallback: ((Int) -> Void)?

    var finishCallback: ((CGFloat, Data) -> Void)?

    var audioDuration: CGFloat {
        guard !mp3Path.isEmpty else { return .zero }
        return CMTimeGetSeconds(AVURLAsset(url: URL(fileURLWithPath: mp3Path)).duration)
    }
}

private extension CLRecorder {
    private func initFolder() {
        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: CLPath.Folder.tmp) {
            try? fileManager.createDirectory(atPath: CLPath.Folder.tmp, withIntermediateDirectories: true, attributes: nil)
        }
    }

    private func initRemoteIO() {
        initAudioComponent()
        initFormat()
        initAudioProperty()
        initRecordCallback()
        initAudioUnit()
        initBuffer()
    }

    private func initAudioComponent() {
        var audioDesc = AudioComponentDescription(componentType: kAudioUnitType_Output,
                                                  componentSubType: kAudioUnitSubType_RemoteIO,
                                                  componentManufacturer: kAudioUnitManufacturer_Apple,
                                                  componentFlags: 0,
                                                  componentFlagsMask: 0)

        let inputComponent = AudioComponentFindNext(nil, &audioDesc)
        AudioComponentInstanceNew(inputComponent!, &audioUnit)
    }

    private func initFormat() {
        guard let audioUnit else { return }
        var audioFormat = AudioStreamBasicDescription(mSampleRate: 44100,
                                                      mFormatID: kAudioFormatLinearPCM,
                                                      mFormatFlags: kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsNonInterleaved | kAudioFormatFlagIsPacked,
                                                      mBytesPerPacket: 2,
                                                      mFramesPerPacket: 1,
                                                      mBytesPerFrame: 2,
                                                      mChannelsPerFrame: 1,
                                                      mBitsPerChannel: 16,
                                                      mReserved: 0)

        AudioUnitSetProperty(audioUnit,
                             kAudioUnitProperty_StreamFormat,
                             kAudioUnitScope_Output,
                             1,
                             &audioFormat,
                             UInt32(MemoryLayout<AudioStreamBasicDescription>.size))
    }

    private func initAudioProperty() {
        guard let audioUnit else { return }
        var flag: UInt32 = 1
        AudioUnitSetProperty(audioUnit,
                             kAudioOutputUnitProperty_EnableIO,
                             kAudioUnitScope_Input,
                             1,
                             &flag,
                             UInt32(MemoryLayout<UInt32>.size))
    }

    private func initRecordCallback() {
        guard let audioUnit else { return }

        var recordCallback = AURenderCallbackStruct(inputProc: { inRefCon, ioActionFlags, inTimeStamp, inBusNumber, inNumberFrames, ioData -> OSStatus in
            let recorder = Unmanaged<CLRecorder>.fromOpaque(inRefCon).takeUnretainedValue()
            return recorder.recordCallback(ioActionFlags: ioActionFlags, inTimeStamp: inTimeStamp, inBusNumber: inBusNumber, inNumberFrames: inNumberFrames)
        }, inputProcRefCon: UnsafeMutableRawPointer(Unmanaged.passUnretained(self).toOpaque()))

        AudioUnitSetProperty(audioUnit,
                             kAudioOutputUnitProperty_SetInputCallback,
                             kAudioUnitScope_Global,
                             1,
                             &recordCallback,
                             UInt32(MemoryLayout<AURenderCallbackStruct>.size))
    }

    private func initAudioUnit() {
        guard let audioUnit else { return }
        AudioUnitInitialize(audioUnit)
    }

    private func initBuffer() {
        guard let audioUnit else { return }
        var flag: UInt32 = 0
        AudioUnitSetProperty(audioUnit,
                             kAudioUnitProperty_ShouldAllocateBuffer,
                             kAudioUnitScope_Output,
                             1,
                             &flag,
                             UInt32(MemoryLayout<UInt32>.size))
    }

    private func recordCallback(ioActionFlags: UnsafeMutablePointer<AudioUnitRenderActionFlags>,
                                inTimeStamp: UnsafePointer<AudioTimeStamp>,
                                inBusNumber: UInt32,
                                inNumberFrames: UInt32) -> OSStatus
    {
        var bufferList = AudioBufferList(mNumberBuffers: 1,
                                         mBuffers: AudioBuffer(mNumberChannels: 1,
                                                               mDataByteSize: 0,
                                                               mData: nil))
        AudioUnitRender(audioUnit!, ioActionFlags, inTimeStamp, inBusNumber, inNumberFrames, &bufferList)
        mp3Encoder.processAudioBufferList(bufferList)
        let second = Int(floor(audioDuration))
        if lastSecond != second {
            lastSecond = second
            DispatchQueue.main.async {
                self.durationCallback?(second)
            }
        }
        return noErr
    }

    private func removeFile() {
        DispatchQueue.global().async {
            try? FileManager.default.removeItem(atPath: self.mp3Path)
            self.mp3Path = ""
        }
    }

    private func activateAudioSession() {
        let audioSession = AVAudioSession.sharedInstance()
        otherAudioPlaying = audioSession.isOtherAudioPlaying
        try? audioSession.setCategory(.playAndRecord, options: .defaultToSpeaker)
        try? audioSession.setPreferredSampleRate(44100)
        try? audioSession.setPreferredInputNumberOfChannels(1)
        try? audioSession.setPreferredIOBufferDuration(0.023)
        try? audioSession.setActive(true)
    }

    private func resumeAudioSession() {
        let audioSession = AVAudioSession.sharedInstance()
        if otherAudioPlaying {
            try? audioSession.setActive(true, options: .notifyOthersOnDeactivation)
        } else {
            try? audioSession.setCategory(.playback)
            try? audioSession.setActive(true)
        }
    }

    private func writeData(_ data: Data) {
        lock.lock()
        fileHandle?.seekToEndOfFile()
        fileHandle?.write(data)
        lock.unlock()
    }
}

extension CLRecorder {
    func start() {
        guard let audioUnit else { return }
        lastSecond = 0
        DispatchQueue.main.async {
            self.durationCallback?(0)
        }

        mp3Path = CLPath.Folder.chatFile.appending("/\(Int(Date().timeIntervalSince1970 * 1_000_000)).mp3")
        if !FileManager.default.fileExists(atPath: mp3Path) {
            FileManager.default.createFile(atPath: mp3Path, contents: nil, attributes: nil)
        }

        fileHandle = try? FileHandle(forWritingTo: URL(fileURLWithPath: mp3Path))
        if fileHandle != nil {
            activateAudioSession()
            let status = AudioOutputUnitStart(audioUnit)
            if status != noErr {
                CLLog("startRecorder error \(status)")
            }
        } else {
            CLLog("error: unable to create file handle")
        }
    }

    func cancel() {
        guard let audioUnit else { return }

        let status = AudioOutputUnitStop(audioUnit)
        if status != noErr {
            CLLog("stopRecorder error: \(status)")
        }
        resumeAudioSession()
        removeFile()
    }

    func stop() {
        guard let audioUnit else { return }

        let status = AudioOutputUnitStop(audioUnit)
        resumeAudioSession()
        if status != noErr {
            CLLog("stopRecorder error: \(status)")
        } else {
            guard let fileData = try? Data(contentsOf: URL(fileURLWithPath: mp3Path)) else { return }
            DispatchQueue.main.async {
                self.finishCallback?(self.audioDuration, fileData)
                self.removeFile()
            }
        }
    }
}
